### PR TITLE
Base64 encoding and decoding

### PR DESCRIPTION
**Pull request by coderSomya**
This PR adds a library for base64 encoding and decoding with custom table support for safe url parsing.
Added two internal functions.
- **_encode**: Encodes `bytes` data into a base64-like string using a custom encoding table, with optional padding.
- **_decode**: Decodes a base64-like string back to `bytes` using a custom decoding table, handling padding.

to test
```
forge test
```

### CODE REVIEW

File: test/TestBase64.t.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    using Base64 for bytes;
+
+    function testEncode() public {
+        bytes memory data = bytes("hello world");
+        string memory encoded = Base64.encode(data);
+        string memory expected = "aGVsbG8gd29ybGQ="; // "hello world" in Base64
+        assertEq(encoded, expected, "Base64 encode failed");
+    }
+
+    function testEncodeURL() public {
```
Comment (pablin-dev) [inline test/TestBase64.t.sol:17]: URL test should contain `file` or `urls`. For example:
- URL=`"http://pepe.com:8545"`; // aHR0cDovL3BlcGUuY29tOjg1NDU= ;  
- URL=`file://pepe.json` ZmlsZTovL3BlcGUuanNvbg==

File: test/TestBase64.t.sol [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    function testEncodeDecode() public {
+        bytes memory data = bytes("Test encoding and decoding round-trip!");
+        string memory encoded = Base64.encode(data);
+        bytes memory decoded = Base64.decode(encoded);
+        assertEq(decoded, data, "Round-trip Base64 encode/decode failed");
+    }
+
+    function testEncodeDecodeURL() public {
+        bytes memory data = bytes("Test encoding and decoding round-trip URL!");
```
Comment (pablin-dev) [inline test/TestBase64.t.sol:37]: Same as previous.

File: contracts/utils/Base64.sol [hunk 3]
```diff
@@ -4,27 +4,134 @@ pragma solidity ^0.8.24;
...
+     * @dev Internal encoding function supporting table lookup and optional padding.
+     * This function encodes a given `data` (in `bytes` format) into a string using a custom base64-like encoding table.
+     * Padding is optional based on the `withPadding` argument.
+     * 
+     * @param data The data to be encoded in bytes format.
+     * @param table The table to be used for encoding (typically a string of characters representing the base64 alphabet).
+     * @param withPadding A boolean indicating whether padding should be added to the result.
+     * @return result The encoded string.
+     */
+    function _encode(bytes memory data, string memory table, bool withPadding) private pure returns (string memory) {
```
Comment (pablin-dev) [inline contracts/utils/Base64.sol:42]: - Is there a `max` lenght for `data` input on `_decode()` and `_encode()`?

**Review by pablin-dev — CHANGES_REQUESTED** [review]

**Comment by pablin-dev** [comment]
One more question: did `bun` have any issue with the JavaScript side? I notice you change it in your PR for `yarn`.

File: contracts/utils/Base64.sol (hunk 3 above)
Comment (coderSomya) [reply contracts/utils/Base64.sol:42]: it will be good to have it. will see to it. thanks

**Comment by coderSomya** [comment]
> One more question: did `bun` have any issue with the JavaScript side? I notice you change it in your PR for `yarn`.

No issues with bun, i've been using that itself for package management. My change is only a spelling correction in package.json besides the base64 stuff.

**Review by pablin-dev — APPROVED** [review]
LGTM! Good work! You can merge it.

**Pull request merged**